src/Components/ServiceScene/Breakdowns/LineFilterInput.tsx

```diff
@@ -1,41 +1,110 @@
 import { css } from '@emotion/css';
-import { Icon, IconButton, Input, useStyles2 } from '@grafana/ui';
-import React, { HTMLProps } from 'react';
+import { Icon, IconButton, Input, Tooltip, useStyles2 } from '@grafana/ui';
+import React, { ChangeEvent, HTMLProps, useCallback, useEffect, useState } from 'react';
 import { GrafanaTheme2 } from '@grafana/data';
+import { narrowErrorMessage } from '../../../services/narrowing';
 
-interface Props extends Omit<HTMLProps<HTMLInputElement>, 'width' | 'prefix'> {
+interface Props extends Omit<HTMLProps<HTMLInputElement>, 'width' | 'prefix' | 'invalid' | 'onInvalid'> {
   onClear?: () => void;
   suffix?: React.ReactNode;
   prefix?: React.ReactNode;
   width?: number;
+  regex: boolean;
+  value: string;
 }
 
-export const LineFilterInput = ({ value, onChange, placeholder, onClear, suffix, width, ...rest }: Props) => {
+let re2JS: typeof import('re2js').RE2JS | undefined | null = undefined;
+
+export const LineFilterInput = ({ value, onChange, placeholder, onClear, suffix, width, regex, ...rest }: Props) => {
   const styles = useStyles2(getStyles);
-  return (
-    <Input
-      rows={2}
-      width={width}
-      value={value}
-      onChange={onChange}
-      suffix={
-        <span className={styles.suffixWrapper}>
-          {onClear && value ? (
-            <IconButton
-              aria-label={'Clear line filter'}
-              tooltip={'Clear line filter'}
-              onClick={onClear}
-              name="times"
-              className={styles.clearIcon}
-            />
-          ) : undefined}
-          {suffix && suffix}
-        </span>
+  const [invalid, setInvalid] = useState(false);
+  const [errorMessage, setErrorMessage] = useState('');
+
+  const validate = useCallback(
+    (value: string) => {
+      if (value) {
+        try {
+          re2JS?.compile(value);
+          if (invalid) {
+            setInvalid(false);
+            setErrorMessage('');
+          }
+        } catch (e) {
+          const msg = narrowErrorMessage(e);
+          if (!invalid) {
+            setInvalid(true);
+          }
+
+          if (msg && msg !== errorMessage) {
+            setErrorMessage(msg);
+          }
+        }
+      } else if (invalid) {
+        setInvalid(false);
+        setErrorMessage('');
       }
-      prefix={<Icon name="search" />}
-      placeholder={placeholder}
-      {...rest}
-    />
+    },
+    [errorMessage, invalid]
+  );
+
+  const load = useCallback(async () => {
+    re2JS = null;
+    re2JS = (await import('re2js')).RE2JS;
+  }, []);
+
+  const initValidation = useCallback(
+    (value: string) => {
+      if (re2JS === undefined && regex) {
+        load().then(() => validate(value));
+      } else if (regex && re2JS) {
+        validate(value);
+      }
+    },
+    [load, regex, validate]
+  );
+
+  useEffect(() => {
+    if (!regex) {
+      setInvalid(false);
+      setErrorMessage('');
+    }
+  }, [regex]);
+
+  initValidation(value);
```

Comment:
Do you need this to run on every re-render?

Reply:
nah we can wrap in a useEffect and run on init, good catch

Reply:
Yeah, better, becuse it might be running on every keystroke.

Reply:
Should be cleaned in https://github.com/grafana/logs-drilldown/pull/1190/commits/797a27fe9814a797034c2068959f31461edb586f, it'll still run once on every keystroke, but not twice!